### PR TITLE
SideNavigation: V2

### PR DESCRIPTION
**Pull request by AlbertCarreras**
#### What changed?

SideNavigation: V2 release including nested directories

#### Why?

V2 in alpha release. More changes coming to the component including improved keyboard navigation 

Breaking changes: 

- SideNavigation.Item API changes with supporting codemods for manual update

 - `onSelected` to `onClick` 

```
yarn codemod detectManualReplacement ~/path/to/your/code
--component=SideNavigation
--subcomponent=Item
--prop=onSelected
```

 - `item` to `label`
 
```
yarn codemod detectManualReplacement ~/path/to/your/code
--component=SideNavigation
--subcomponent=Item
--prop=item
```

 - `SideNavigation.Item` to `SideNavigation.TopItem`
 
```
yarn codemod detectManualReplacement ~/path/to/your/code
--component=SideNavigation
--subcomponent=Item
--prop=href
```

### Links

- [Figma](https://www.figma.com/file/kK9TSF41utPLF2CIRoGLhX/Gestalt-SideNav-Component?node-id=285%3A4844)
- [TDD](https://paper.dropbox.com/doc/TDD-SideNavigation--BkRWqAZLEiLaYaoK1dA32uFvAg-4IWmy1gXM95tF7SNNKh2A)

### Checklist

- [x] Added unit and Flow Tests
- [x] Added documentation + accessibility tests
- [x] Verified accessibility: keyboard & screen reader interaction
- [x] Checked dark mode, responsiveness, and right-to-left support
- [x] Checked stakeholder feedback (e.g. Gestalt designers)

### CODE REVIEW

File: packages/gestalt/src/SideNavigationTopItem.js [hunk 1]
```diff
@@ -41,36 +49,38 @@ type Props = {|
...
   icon,
+  label,
   notificationAccessibilityLabel,
-  onSelect,
+  onClick,
 }: Props): Node {
+  const { nestedLevel } = useNesting();
+  const { setSelectedItemId } = useSideNavigation();
+
+  const itemId = useId();
```
Comment (ayeshakmaz) [inline packages/gestalt/src/SideNavigationTopItem.js:82]: oooo nice!

File: packages/gestalt/src/useGetChildrenToArray.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+};
+
+const useGetChildrenToArray = ({
+  children,
+  filterLevel,
+}: {|
+  children: Node,
+  filterLevel: 'main' | 'nested',
+|}): $ReadOnlyArray<Node> => {
+  let hasFirstNavigationItem = false;
```
Comment (ayeshakmaz) [inline packages/gestalt/src/useGetChildrenToArray.js:16]: I think this var name is not the clearest, is it more like "foundFirstNavigationItem"?

File: packages/gestalt/src/SideNavigationGroup.js [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+            >
+              <Text align="end" color="default">
+                {counter.number}
+              </Text>
+            </Box>
+          </Flex.Item>
+        )}
+        {display === 'expandable' ? (
+          <Flex.Item flex="none" alignSelf="center">
+            {/* marginEnd={-2} is a hack to correctly position the counter as Flex + gap + width="100%" doean't expand to full width */}
```
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply packages/gestalt/src/SideNavigationGroup.js:123]: I tried a lot here. I put together the Flex training so I had a good understanding of the existing possibilities and caveats. The issue here is that when I remove marginEnd={-2} and remove width="100%" from the parent to get the correct spacing then another design spec gets disaligned. Then flex="grow" breaks. And it breaks because we are setting a height in Box that's needed to self center the contents. 

I'd ship like this and maybe revisit it with fresh eyes. I spent an entire evening and finally marginEnd={-2}  came to my mind as the only salvation.

File: packages/gestalt/src/SideNavigationGroup.js [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+  const { selectedItemId, setSelectedItemId } = useSideNavigation();
+
+  const navigationChildren = useGetChildrenToArray({
+    children,
+    filterLevel: 'nested',
+  });
+
+  let itemColor;
+
+  if (hovered && !expanded) {
```
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply packages/gestalt/src/SideNavigationGroup.js:192]: @hectoid Thoughts?

File: packages/gestalt/src/SideNavigationGroup.js [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+                ...paddingStyle,
+              }
+            : paddingStyle,
+      }}
+    >
+      <Flex gap={2} height="100%" width="100%">
+        {icon ? (
+          <Flex.Item alignSelf="center">
+            <Box aria-hidden>
+              {typeof icon === 'string' ? (
```
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply packages/gestalt/src/SideNavigationGroup.js:75]: nope, custom icons are objects {{ __path }}

File: packages/gestalt/src/SideNavigationGroup.js [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+  const paddingStyle = {
+    paddingInlineStart: NESTING_MARGIN_START_MAP[nestedLevel],
+    paddingInlineEnd: '16px',
+  };
+
+  return (
+    <li className={classnames(styles.liItem)}>
+      <NestingProvider>
+        {display === 'expandable' ? (
+          <TapArea
```
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply packages/gestalt/src/SideNavigationGroup.js:205]: True! missed it

File: packages/gestalt/src/SideNavigationTopItem.js (hunk 1 above)
Comment (AlbertCarreras) [reply packages/gestalt/src/SideNavigationTopItem.js:82]: very nice

File: packages/gestalt/src/useGetChildrenToArray.js (hunk 2 above)
Comment (AlbertCarreras) [reply packages/gestalt/src/useGetChildrenToArray.js:16]: DONE

File: packages/gestalt/src/SideNavigationGroup.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (ayeshakmaz) [reply packages/gestalt/src/SideNavigationGroup.js:123]: hahah I trust you, was just curious cuz I noticed the same on SlimBanner - this looks good!

File: packages/gestalt/src/SideNavigationGroup.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (ayeshakmaz) [reply packages/gestalt/src/SideNavigationGroup.js:75]: Oooh I see it now, ok cool

**Review by ayeshakmaz — APPROVED** [review]

File: packages/gestalt/src/SideNavigationTopItem.js [hunk 7]
```diff
@@ -41,36 +49,38 @@ type Props = {|
...
-  item: {| label: string, value: string |},
+  label: string,
 |};
 
 /**
- * Use [SideNavigation.Item](https://gestalt.pinterest.systems/sidenavigation#SideNavigation.Item) for navigation, when the SideNavigation item navigates to a new page or sectoin.
+ * Use [SideNavigation.TopItem](https://gestalt.pinterest.systems/sidenavigation#SideNavigation.TopItem) for navigation, when the SideNavigation.TopItem navigates to a new page or sectoin.
  */
-export default function SideNavigationItem({
+export default function SideNavigationTopItem({
```
Comment (dangerismycat) [inline packages/gestalt/src/SideNavigationTopItem.js:69]: Can we keep the name of this component as "Item"? For the simplest use cases, it'd be nice to keep usage as something like this
```js
<SideNav>
  <SideNav.Item />
  <SideNav.Item />
  <SideNav.Item />
</SideNav>
```

File: docs/pages/design.js [hunk 8]
```diff
@@ -6,10 +6,10 @@ import PageHeader from '../components/PageHeader.js';
 
 export default function DocsPage(): Node {
   return (
-    <Page title="Getting started as a designer">
-      <PageHeader name="Getting started as a designer" description="" type="guidelines" />
+    <Page title="Get started as a designer">
```
Comment (dangerismycat) [inline docs/pages/design.js:9]: nit: let's keep PRs focused and move unrelated changes like these out to separate cleanup PRs, especially for these larger PRs that are difficult to review already

File: packages/gestalt/src/SideNavigationNestedItem.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+      | SyntheticKeyboardEvent<HTMLAnchorElement>,
+    dangerouslyDisableOnNavigation: () => void,
+  |}) => void,
+|};
+
+/**
+ * Use [SideNavigation.NestedItem](https://gestalt.pinterest.systems/sidenavigation#SideNavigation.NestedItem)
+ */
+export default function SideNavigationNestedItem({ active, href, label, onClick }: Props): Node {
+  return <SideNavigationTopItem active={active} href={href} label={label} onClick={onClick} />;
```
Comment (dangerismycat) [inline packages/gestalt/src/SideNavigationNestedItem.js:35]: This is confusing. So TopItem is really more of a BaseItem?

File: packages/gestalt/src/contexts/NestingProvider.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+const NestingContext: Context<NestingContextType> = createContext<NestingContextType>({
+  nestedLevel: 0,
+});
+
+const { Provider } = NestingContext;
+
+function NestingProvider({ children }: Props): Element<typeof Provider> {
+  const { nestedLevel } = useContext(NestingContext);
+
+  const nextNestedLevel = (nestedLevel ?? 0) + 1;
```
Comment (dangerismycat) [inline packages/gestalt/src/contexts/NestingProvider.js:21]: If I'm reading this logic correctly, we can only ever have nested levels of `1 | 2`? If `nestedLevel` doesn't exist, we start with 0, add 1 to it, then pass down the result. Can a component ever read the context and get `0`?

File: packages/gestalt/src/useGetChildrenToArray.js [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+  const navigationChildren = [];
+
+  let recursionLevel = 0;
+
+  const getChildren = ({ nodeChildren }) =>
+    Children.toArray(nodeChildren).forEach((child) => {
+      // Detect incorrect subcomponent usage
+      if (filterLevel === 'nested' && ALLOWED_CHILDREN_MAP.main.includes(child.type.displayName)) {
```
Comment (dangerismycat) [inline packages/gestalt/src/useGetChildrenToArray.js:25]: These two checks can be combined:
```js
if (!ALLOWED_CHILDREN_MAP[filterLevel].includes(child.type.displayName) { throw error }
```

File: packages/gestalt/src/useGetChildrenToArray.js [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+          navigationChildren.push(cloneElement(child, { _hasMarginTop: true }));
+        }
+      } else if (ALLOWED_CHILDREN_MAP[filterLevel].includes(child.type.displayName)) {
+        if (!hasFirstNavigationItem) {
+          hasFirstNavigationItem = true;
+        }
+        navigationChildren.push(child);
+      } else if (!child?.type?.displayName?.startsWith('SideNavigation')) {
+        recursionLevel += 1;
+        if (recursionLevel < 2) {
```
Comment (dangerismycat) [inline packages/gestalt/src/useGetChildrenToArray.js:48]: This should also throw an error if `recursionLevel >= 2`, otherwise it will fail silently — yes, the Context currently throws in that case, but that could get refactored in the future, leading to unexpected behavior here
